mono/metadata/threads.c

```diff
@@ -1753,39 +1753,56 @@ ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj,
 
 /* If the array is already in the requested domain, we just return it,
    otherwise we return a copy in that domain. */
-static MonoArray*
-byte_array_to_domain (MonoArray *arr, MonoDomain *domain, MonoError *error)
+static MonoArrayHandle
+byte_array_to_domain (MonoArrayHandle arr, MonoDomain *domain, MonoError *error)
 {
-	MonoArray *copy;
+	HANDLE_FUNCTION_ENTER ()
 
-	error_init (error);
-	if (!arr)
-		return NULL;
+	if (!arr || MONO_HANDLE_IS_NULL (arr))
+		return MONO_HANDLE_NEW (MonoArray, NULL);
 
-	if (mono_object_domain (arr) == domain)
+	if (MONO_HANDLE_DOMAIN (arr) == domain)
 		return arr;
 
-	copy = mono_array_new_checked (domain, mono_defaults.byte_class, arr->max_length, error);
-	memmove (mono_array_addr (copy, guint8, 0), mono_array_addr (arr, guint8, 0), arr->max_length);
-	return copy;
+	size_t const size = mono_array_handle_length (arr);
+
+	// Capture arrays into array for looping.
+	struct {
+		MonoArrayHandle handle;
+		gpointer p;
+		guint gchandle;
+	} arrays [] = {
+		{ arr },
+		{ mono_array_new_handle (domain, mono_defaults.byte_class, size, error) },
+	};
+	goto_if_nok (error, exit);
+
+	guint i;
+
+	// Pin both arrays.
+	for (i = 0; i <= 1; ++i)
+		arrays [i].p = mono_array_handle_pin_with_size (arrays [i].handle, size, 0, &arrays [i].gchandle);
+	goto_if_nok (error, exit);
+
+	memmove (arrays [1].p, arrays [0].p, size);
```

Comment:
Maybe promote to a function like `mono_array_handle_memmove_norefs (MonoArrayHandle dest, guint32 dest_idx, MonoArrayHandle src, guint32 src_idx, guint32 nelems)` that does the pinning  and memmove?  Might be useful to have in the future.

Reply:
Ugh I keep losing changes on this one and the array is back.